BUGFIX - Modifie la méthode clean du modèle Cats pour autoriser les noms de chats identiques par utilisateur

iglesias1602 opened the pull request:
Dans le programme il y avait un comportément innatendu où deux utilisateurs ne peuvent pas avoir des chats avec le même nom, exemple:

- *User1* crée un chat "Toto" dans son compte.
- *User2* essaye de créer un chat "Toto" dans son compte aussi, mais le programme lui dit qu'il y a eu un erreur.

Ce commit ajuste la logique de la méthode clean dans le modèle Cats. Ça a été fait en filtrant les chats par nom et par identifiant du propriétaire (ownerId).

Avant, cette méthode empêchait la création de chats avec des noms identiques dans l'ensemble de la base de données. Avec ce changement, elle permet désormais à différents utilisateurs d'avoir des chats portant le même nom, tout en empêchant un utilisateur individuel d'avoir plusieurs chats avec le même nom.

Cette mise à jour garantit une plus grande flexibilité pour nommer des chats tout en maintenant l'intégrité des données.

iglesias1602 committed:
BUGFIX - Modifie la méthode clean du modèle Cats pour autoriser les noms de chats identiques par utilisateur

Dans le programme il y avait un comportément innatendu où deux
utilisateurs ne peuvent pas avoir des chats avec le même nom, exemple:

User1 crée un chat "Toto" dans son compte.
User2 essaye de créer un chat "Toto" dans son compte aussi, mais le programme
lui dit qu'il y a eu un erreur.

Ce commit ajuste la logique de la méthode clean dans le modèle Cats.
Avant, cette méthode empêchait la création de chats avec des noms identiques dans l'ensemble de la base de données.
Avec ce changement, elle permet désormais à différents utilisateurs d'avoir des chats portant le même nom, tout en empêchant
un utilisateur individuel d'avoir plusieurs chats avec le même nom.

Ça a été réalisé en filtrant les chats par nom et par identifiant du propriétaire (ownerId) dans la méthode clean.

Cette mise à jour garantit une plus grande flexibilité pour nommer des chats tout en maintenant l'intégrité des données.

iglesias1602 commented:
**Petite précision:** Les failing checks ne sont pas liés à ma PR et ces tests*

PirLucas commented on CICO/models.py:
```diff
@@ -69,8 +69,9 @@ class Cats(models.Model):
     string_value5 = models.CharField(max_length=200, blank=True)
 
     def clean(self):
-        if Cats.objects.filter(name=self.name).exists():
+        if Cats.objects.filter(name=self.name, ownerId=self.ownerId_id).exclude(catId=self.catId).exists():
             raise ValidationError("A cat with this name already exists for this user.")
```
Est-ce nécessaire d'exclure l'id du chat actuel dans la recherche? Puisque le chat n'est normalement pas encore dans la DB.

PirLucas commented on CICO/tests.py:
```diff
@@ -165,6 +165,41 @@ def test_add_cat_unauthorized(self):
+
+    def add_cat(self, cat_name):
+        """ Helper method to add a cat with the given name """
+        img_url = 'https://images5.alphacoders.com/325/thumb-1920-325672.jpg'
```
D'où viens cette image? Est-ce une sorte de base de donnée en ligne d'images stocks? Il faudra prévoir des images enregistrées en local pour faire ce genre de tests, car c'est risqué de se reposer sur une source tierce.

PirLucas commented:
Sinon rien à dire pour le fonctionnement, ça marche comme décrit.

iglesias1602 committed:
Fix des tests de fonctionnalité d'Arthur

*TL;DR:*
*Correction des tests obsolètes, utilisation de 'Mock' pour simuler le temps actuel, simplification de la manière dont on obtient le temps et dans la View*

J'ai fait une correction dans les tests de la classe 'TestsGraphiquesDesChats'. J'avais des vieux tests qui ne fonctionnaient plus, et qui faisaient planter
les tests de fonctionnalité.

Tout d'abord j'ai pris les tests qui avaient été mis à jour dans la branche d'Arthur.

Comme le modèle DeviceRecords met le temps actuel automatiquement ```"time = models.DateTimeField(auto_now_add=True)"```
et pas celui qui est défini lors de tests, il faut utiliser la librairie 'Mock' pour simuler le temps courant.

J'ai utilisé aussi une seule libraire pour obtenir le temps courant, et pas deux.
Avant on utilisait 'timezone.localtime()' et 'timezone.now()'. En vue que les deux méthodes renvoient le temps mais
pas au même format, j'ai utilisé 'timezone.now()'. timezone.now() utilise le format 'UTC' tandis que 'timezone.localtime()' utilise
le format 'ETC', ce qui a posé des problèmes lors de la génération des Records pour les tests unitaires.

J'ai aussi modifié une ligne de code dans la view qui gère le temps du début ```start_date = end_date - timedelta(days=7)``` pour avoir correctement les derniers 7 jours (le jour actuel inclus).

PirLucas reviewed (commented):
,m:;,:
